Create Blog “using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application”

verdog opened the pull request:
Automatically generated by Netlify CMS

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+
+This is where gdb4hpc comes in. gdb4hpc is part of the HPE Cray Programming Environment package. It captures the powerful features of gdb and allows the user to apply them to an HPC application at scale. gdb4hpc can debug HPC applications running on thousands of nodes.
+
+gdb4hpc works by connecting the user to many instances of gdb at once. gdb4hpc controls all of the instances of gdb at once and aggregates and filters the results into representations that will comfortably fit on a single terminal screen.
```
And it would really be much better to add words around it when you are referring to it at the beginning of a paragraph. Words like "tool", "debugger", "app", whatever it actually is, will help the reader have a better picture of it in their mind than just use the name of the tool. In the second sentence, you can use a pronoun, like "It controls all of the instances..."

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+
+## Prerequisites
+
+This tutorial assumes you already have a familiarity with command line debuggers, git, MPI, and CUDA.
```
"git" should be written as "Git"

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+To set up the debugging session, you will need to obtain the sample application
+and compile it for your chosen GPU with debug information.
+
+### Download The Sample Code
```
Brand requests that we use "sentence case" as opposed to "title case". So, the header should read "Download the sample code"

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+
+For this tutorial, you will use the `simpleMPI` example from
+[NVIDIA's cuda-samples repository](https://github.com/NVIDIA/cuda-samples).
+`simpleMPI` does a simple distributed calculation involving square roots and averages:
```
I should say the same thing for "simpleMPI", but since it's in a code block, visually it kind of works this way.

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
````diff
@@ -0,0 +1,545 @@
+simpleMPI.cpp  simpleMPI.cu  simpleMPI.h
+```
+
+Remove the other files if desired.
````
Place a comma after "files".

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
````diff
@@ -0,0 +1,545 @@
+$ rm *.sln *.vcxproj README.md Makefile
+```
+
+### Build The Application
````
Sentence case means this should read "Build the application"

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+To compile the application, load the appropriate compilers and compile the
+sample application for your chosen GPU architecture with debug information.
+
+#### Load Compilers
```
Sentence case would have us write this as "Load compilers". If you think about it, it really changes the meaning. When you write it as "Load Compilers", "Load" appears to be an adjective to the plural noun "Compilers". When you say "Load compilers", you are giving an instruction or command.

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+
+When writing this tutorial, I used CCE 17 and CUDA 12 and ran the application on NVIDIA A100 GPUs, but the debugging process is the same for any compiler combination.
+
+After deciding which versions you want, use the `module` command in a terminal to ensure that the Cray `CC` compiler and the NVIDIA `nvcc` compiler are available.
```
Would the word "need" work better than "want" in that first phrase?

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
````diff
@@ -0,0 +1,545 @@
+...
+```
+
+#### Compile the Application
````
"Compile the application"

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+The sample application comes with a Makefile, but it's not compatible with the
+HPE Cray Programming Environment. You will compile the application manually.
+
+#### Debug Info
```
"Debug info"

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+the compiler to disable these optimizations and include extra information that
+the debugger can use to reason about the original code.
+
+#### Compile the .cu File
```
"File" should be "file"

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+an executable. After you compile the .cpp file in the next section, you will
+link the object files together to create the final executable.
+
+nvcc does not actually fully compile C/C++. It actually processes the CUDA
```
This would be another great place to say "The nvcc [something] does not actually" so as to not start the sentence and paragraph with a lower case letter. (I'm not exactly sure what to call it, though. I'll leave that up to you.)

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+and the [compilation section of the cuda-gdb documentation](https://docs.nvidia.com/cuda/cuda-gdb/index.html#compiling-the-application)
+for more information.
+
+#### Compile the .cpp File
```
Change "File" to "file"

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+The flags have the same meanings as in the nvcc compilation. The Cray CC
+compiler will automatically include the flags needed for building with MPI.
+
+#### Link the Application and Produce an Executable
```
"Link the application and produce an executable"

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
````diff
@@ -0,0 +1,545 @@
+$ CC simpleMPI_gpu.o simpleMPI_cpu.o -o simpleMPI
+```
+
+#### Test The Application
````
"Test the application" (even if you retained title case, "the" would always have a lower case "t")

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+
+`--exclusive` tells Slurm to run on nodes that aren't being used by anyone
+else, and to not let anyone else use them while running the job. Only one
+application can use each GPU at a time, so we need to make sure that we aren't
```
I recommend to our writers to avoid the "Royal we". Stick with the pronouns you've used earlier. You've been giving instructions to "you", and that is perfect. In this case, you would just say "...so you need to make sure that you aren't given codes that have GPUs that are already in use."

verdog replied:
Thanks, this tutorial was actually converted from a previously written tutorial that used the royal we, and I thought I caught all the instances, hence the correctness up to this point. 😅

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+
+In my specific example:
+
+`-n8 -N2 --ntasks-per-node 4` tells Slurm to run 8 ranks across 2 nodes,
```
I'd recommend modifying the last sentence here to "This example uses nodes with 4 A100s each." (see my note below on the use of the "Royal we")

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+For this tutorial, I am using gdb4hpc 4.16.3, but any version above 4.13.1
+will have the commands used in this tutorial.
+
+gdb4hpc will use the cuda-gdb debugger for GPU debugging. cuda-gdb is
```
Please fix the beginning of the sentence using my advice from earlier.

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
````diff
@@ -0,0 +1,545 @@
+There is NO WARRANTY, to the extent permitted by law.
+```
+
+#### Launch the Application in gdb4hpc
````
"Launch the application in gdb4hpc"

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+
+After the launch completes, gdb4hpc stops at `main` and gives you control.
+
+#### Stop at a Breakpoint in a GPU kernel
```
"Stop at a breakpoint in a GPU kernel"

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+
+The application has stopped in a GPU kernel. From here you can use the usual
+debugger commands like `print`, `step`, `next`, etc. You can use `print` to
+inspect GPU memory as you would any other memory. Here we print the first 8 items
```
Modify this sentence that currently reads "Here we print the first 8 items..." so you are not using the Royal We. Try this instead: "Print the first 8 items of the input array."

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+
+#### CUDA Threads and CUDA Commands
+
+gdb4hpc supports cuda-gdb's `cuda` commands.
```
Please change per my earlier comments.

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
````diff
@@ -0,0 +1,545 @@
+simpleMPI{7}: {0.839492,0.778462,0.195212,0.832727,0.125331,0.211325,0.0624168,0.56848}
+```
+
+#### CUDA Threads and CUDA Commands
````
"CUDA threads and CUDA commands"

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
````diff
@@ -0,0 +1,545 @@
+simpleMPI{0..7}: {x = 0, y = 0, z = 0}
+```
+
+#### More CUDA Commands
````
"More CUDA commands"

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+
+See [the cuda-gdb documentation](https://docs.nvidia.com/cuda/cuda-gdb/index.html) for more information.
+
+#### Single Stepping
```
"Single stepping"

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
````diff
@@ -0,0 +1,545 @@
+simpleMPI{7}: 0.646477
+```
+
+## Next Steps
````
"Next steps"

drensing commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+
+## Next Steps
+
+You now know how to start debugging a multinode MPI/CUDA application with gdb4hpc in the HPE Cray Programming Environment!
```
Is there any other benefit statement you can add? Like "This will save you hours of time in trying to find those little bugs."?

drensing reviewed (changes requested):
Very nicely written, Josh! Just a few things I pointed out:
1) try to not start a sentence or paragraph with a lower-case letter
2) Avoid the "Royal We" - there were only a few places where I found this (I may have missed some, so you might want to do a global search and see if you find any more). In generally, you did a great job of avoiding it.
3) Remember to use sentence case, even for headers.

All the best,
Dale

drensing commented:
[like]  Rensing, Dale reacted to your message:
________________________________
From: Josh ***@***.***>
Sent: Thursday, January 23, 2025 5:43:35 PM
To: hpe-dev-incubator/hpe-dev-portal ***@***.***>
Cc: Rensing, Dale ***@***.***>; Review requested ***@***.***>
Subject: Re: [hpe-dev-incubator/hpe-dev-portal] Create Blog “using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application” (PR #2877)


@verdog commented on this pull request.

________________________________

In content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md<https://github.com/hpe-dev-incubator/hpe-dev-portal/pull/2877#discussion_r1927412371>:

> +```
+
+Your srun command's flags may vary depending on your system. You should launch the
+job on nodes with your targeted GPU with one rank per GPU. See
+[the `srun` documentation](https://slurm.schedmd.com/srun.html)
+for how to achieve a correct job launch on your system.
+
+In my specific example:
+
+`-n8 -N2 --ntasks-per-node 4` tells Slurm to run 8 ranks across 2 nodes,
+splitting the job evenly by running 4 ranks on each node. In this example we
+are using nodes with 4 A100s each.
+
+`--exclusive` tells Slurm to run on nodes that aren't being used by anyone
+else, and to not let anyone else use them while running the job. Only one
+application can use each GPU at a time, so we need to make sure that we aren't


Thanks, this tutorial was actually converted from a previously written tutorial that used the royal we, and I thought I caught all the instances, hence the correctness up to this point. 😅

—
Reply to this email directly, view it on GitHub<https://github.com/hpe-dev-incubator/hpe-dev-portal/pull/2877#discussion_r1927412371>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ASPH7RK7GIZWUI4PSXPCQLL2MES4PAVCNFSM6AAAAABVMY2BOCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKNZQGU4TAMJQGU>.
You are receiving this because your review was requested.Message ID: ***@***.***>

verdog commented on content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md:
```diff
@@ -0,0 +1,545 @@
+
+In this tutorial, you will learn how to debug a multinode MPI/CUDA application
+with gdb4hpc in the HPE Cray Programming Environment. This tutorial uses a CUDA
+application and NVIDIA GPUs as examples, but the concepts are applicable to HIP
```
In this case, I think "HIP" is more meaningful than the spelled out "Heterogeneous-Compute Interface for Portability" (similar to how "CUDA" is more meaningful than "Compute Unified Device Architecture")

To help clarify I added "HIP (AMD's equivalent of NVIDIA's CUDA)". Let me know what you think!

verdog commented:
Thanks for the reviews! I have:

- updated introduction paragraphs to avoid starting a sentence with a lowercase letter (@drensing)
  - I didn't thouroughly search the document for this, only the instances that were pointed out.
    Let me know if I missed any.
- expanded and revised introduction of gdb4hpc to be more explanatory and appeal to a wider audience (@drensing)
- git -> Git (@drensing)
- Title case -> sentence case (@drensing)
- Minor grammar (@drensing)
- remove royal we (@drensing)
- expand benefit statement (@drensing)
- added hpe-cray-programming-environment tag (@choukden)
- spell out HPC for the first time (@choukden)
- clarify HIP (@choukden)
- add links for Git, MPI, CUDA (@choukden)

drensing commented:
[heart]         Rensing, Dale reacted to your message:
________________________________
From: Josh ***@***.***>
Sent: Thursday, January 23, 2025 6:03:06 PM
To: hpe-dev-incubator/hpe-dev-portal ***@***.***>
Cc: Rensing, Dale ***@***.***>; Mention ***@***.***>
Subject: Re: [hpe-dev-incubator/hpe-dev-portal] Create Blog “using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application” (PR #2877)


Thanks for the reviews! I have:

  *   updated introduction paragraphs to avoid starting a sentence with a lowercase letter ***@***.***<https://github.com/drensing>)
     *   I didn't thouroughly search the document for this, only the instances that were pointed out.
Let me know if I missed any.
  *   expanded and revised introduction of gdb4hpc to be more explanatory and appeal to a wider audience ***@***.***<https://github.com/drensing>)
  *   git -> Git ***@***.***<https://github.com/drensing>)
  *   Title case -> sentence case ***@***.***<https://github.com/drensing>)
  *   Minor grammar ***@***.***<https://github.com/drensing>)
  *   remove royal we ***@***.***<https://github.com/drensing>)
  *   expand benefit statement ***@***.***<https://github.com/drensing>)
  *   added hpe-cray-programming-environment tag ***@***.***<https://github.com/choukden>)
  *   spell out HPC for the first time ***@***.***<https://github.com/choukden>)
  *   clarify HIP ***@***.***<https://github.com/choukden>)
  *   add links for Git, MPI, CUDA ***@***.***<https://github.com/choukden>)

—
Reply to this email directly, view it on GitHub<https://github.com/hpe-dev-incubator/hpe-dev-portal/pull/2877#issuecomment-2610597836>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ASPH7RL62SGAYW6VHT3AKE32MEVFVAVCNFSM6AAAAABVMY2BOCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMJQGU4TOOBTGY>.
You are receiving this because you were mentioned.Message ID: ***@***.***>
